[stdlib] Implementing O(1) methods on `CountableClosedRange`

austinzheng opened the pull request:
#### What's in this pull request?

Functional changes and additional unit test coverage for `CountableRange` and `CountableClosedRange`.

All tests pass on OS X.

@gribozavr 

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

Changes:
- Implemented O(1) `index(_:offsetBy:)` and `distance(from:to:)` methods on `CountableClosedRange`
- Fixed some unit tests
- Added many more unit tests for Countable*Range index APIs

moiseev commented:
@swift-ci Please test

moiseev commented on stdlib/public/core/ClosedRange.swift:
```diff
@@ -47,26 +47,64 @@ public struct ClosedRangeIndex<Bound> : Comparable
+      return 1 + limit.distance(to: right)
+    case (.pastEnd, .pastEnd):
+      // end <--> end
+      return limit.distance(to: limit)
```
`Bound.Stride` conforms to `Integer`, so this can be `0` I guess?

moiseev commented on stdlib/public/core/ClosedRange.swift:
```diff
@@ -47,26 +47,64 @@ public struct ClosedRangeIndex<Bound> : Comparable
+      return 1 + left.distance(to: limit)
+    case let (.pastEnd, .inRange(right)):
+      // in range <-- end
+      return 1 + limit.distance(to: right)
```
The distance form `limit` to `right` is ought to be negative, so it looks like this should actually be something like:
```Swift
return limit.distance(to: right) - 1
```

austinzheng replied:
Oh, good catch. I'll need to make sure the unit tests can check for this condition as well.

moiseev replied:
Yes please. :+1:

Also... Since we're here already, it might make sense to change the requirement to `Bound.Stride : SignedInteger` to reflect this.

moiseev commented:
Thanks @austinzheng for doing this. Please let me know once you've updated the tests and I'll be happy to merge it in.

austinzheng commented:
I also have another PR, whenever you or Dmitri have time: https://github.com/apple/swift/pull/2940

austinzheng commented:
I need to rewrite a bunch of these tests. I'll ping you when I have the updated code ready.

austinzheng commented:
@moiseev

I fixed your comments, addressed some additional bugs, and rewrote some of the test suites I added since they were not properly implemented before. The tests in question pass; I'm running the entire test suite on OS X right now and will comment again when it finishes. Otherwise, it's ready.

austinzheng commented:
All tests pass on my OS X machine.

moiseev commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -412,21 +542,180 @@ ${TestSuite}.test("count/whereBoundIsStrideable") {
+      }
+      previousValue = range[idx].value
+    }
+    // Iterate once more to get to the 'before start' position
```
This does not look right. Past-end is a valid index, before-start is not. Valid indices can only be in the following range [startIndex, endIndex).

austinzheng replied:
Yes. In addition to the comment being misleading, this specific test is really confusingly rewritten and I will redo it to make the intent clearer.

austinzheng commented:
I looked into that test more closely. The test was correct, but written in a very confusing way, so I changed it around and made the preconditions it tests explicit in a comment. The "before start index" comment was simply wrong and I've removed it.

I'm happy to make any other changes you'd like. Also let me know if you want me to squash the commit. @moiseev

moiseev commented:
@swift-ci Please test

austinzheng commented:
Looks like everything's in working order.

gribozavr commented on stdlib/public/core/ClosedRange.swift:
```diff
@@ -47,26 +47,64 @@ public struct ClosedRangeIndex<Bound> : Comparable
   
-  internal func _predecessor(upperBound limit: Bound) -> ClosedRangeIndex {
+  internal func _predecessor(
+    lowerBound: Bound, upperBound limit: Bound
```
At this point, now that you are passing both bounds, I think it would make sense to just move this code into its only callsite.

gribozavr commented on stdlib/public/core/ClosedRange.swift:
```diff
@@ -47,26 +47,64 @@ public struct ClosedRangeIndex<Bound> : Comparable
   internal func _advanced(
-    by n: Bound.Stride, upperBound limit: Bound
+    by n: Bound.Stride, lowerBound: Bound, upperBound limit: Bound
   ) -> ClosedRangeIndex {
```
Same here.

gribozavr commented on stdlib/public/core/ClosedRange.swift:
```diff
@@ -47,26 +47,64 @@ public struct ClosedRangeIndex<Bound> : Comparable
+  }
+
+  internal func _distance(
+    to: ClosedRangeIndex<Bound>, upperBound limit: Bound
```
And here.

moiseev commented:
Ooops. Sorry, @gribozavr, I merged it before seeing your comments.

gribozavr commented:
NP.

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -185,6 +185,136 @@ let clampedTests: [ClampedTest] = [
+
+  /// Return the index `leftIndex` steps away from `start`, or the collection's
+  /// `endIndex` if `leftIndex` is `nil`.
+  func leftIndex<C: RandomAccessCollection>(forCollection c: C) -> C.Index {
```
Please use just `for:` for the label (`Collection` is redundant).

gribozavr replied:
Or even `of:`.

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -185,6 +185,136 @@ let clampedTests: [ClampedTest] = [
+
+  /// Return the index `rightIndex` steps away from `start`, or the
+  /// collection's `endIndex` if `rightIndex` is `nil`.
+  func rightIndex<C: RandomAccessCollection>(forCollection c: C) -> C.Index {
```
Same here.

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -185,6 +185,136 @@ let clampedTests: [ClampedTest] = [
+      return c.endIndex
+    }
+    let iterations = leftIndex - IndexDistanceTest.start
+    _precondition(iterations >= 0)
```
You don't need this precondition, it will be checked by `0..<iterations`.

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -410,21 +540,183 @@ ${TestSuite}.test("count/whereBoundIsStrideable") {
+%     end
+
+  expectCrashLater()
+    _ = range.index(after: idx)
```
Indentation is slightly off here.

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -410,21 +540,183 @@ ${TestSuite}.test("count/whereBoundIsStrideable") {
+    let end = ${Bound}(endValue)
+    let range: ${Self}<${Bound}> = start${op}end
+    var idx = range.startIndex
+    var previousValue : Int? = nil
```
Extra space before the colon.

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -410,21 +540,183 @@ ${TestSuite}.test("count/whereBoundIsStrideable") {
+  var idx = range.startIndex
+
+%     if 'Closed' in Self:
+  // Should be able to advance the range once
```
Did you mean "advance the index"?

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -410,21 +540,183 @@ ${TestSuite}.test("count/whereBoundIsStrideable") {
+    _ = range.index(after: idx)
+}
+
+${TestSuite}.test("index(before:)/semantics") {
```
Same comments as for `index(above:)`.

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -410,21 +540,183 @@ ${TestSuite}.test("count/whereBoundIsStrideable") {
+        expectEqual(newIndex, range[idx].value, stackTrace: SourceLocStack().with(test.loc))
+      }
+    } else {
+      expectCrashLater()
```
This would crash the test and not execute the rest of the tests.  Please use parameterized tests.

gribozavr commented on validation-test/stdlib/Range.swift.gyb:
```diff
@@ -410,21 +540,183 @@ ${TestSuite}.test("count/whereBoundIsStrideable") {
+  let range: ${Self}<${Bound}> = ${Bound}(0)${op}${Bound}(0)
+  var idx = range.startIndex
+
+  // This should work
```
Please add a full stop at the end of comments (here and everywhere).  This helps to distinguish complete comments and ones that the author did not finish but happened to commit anyway.

gribozavr commented:
@austinzheng Great work! 👍
